scipy/stats/_continuous_distns.py

```diff
@@ -2937,6 +2937,202 @@ def _entropy(self, c):
+       and inverse Gaussian distributions: Limiting cases and approximation
+       of processes. FDM Preprint 80, April 2003. University of Freiburg.
+       https://freidok.uni-freiburg.de/fedora/objects/freidok:7974/datastreams/FILE1/content
+
+    %(example)s
+
+    """
+
+    def _argcheck(self, p, a, b):
+        return (np.logical_and(np.abs(b) < a, p >= 0)
+                | np.logical_and(np.abs(b) <= a, p < 0))
+
+    def _logpdf(self, x, p, a, b):
+        # kve instead of kv works better for large values of p
+        # and smaller values of sqrt(a^2  - b^2)
+        @np.vectorize
+        def _logpdf_single(x, p, a, b):
+            return _stats.genhyperbolic_logpdf(x, p, a, b)
+
+        return _logpdf_single(x, p, a, b)
+
+    def _pdf(self, x, p, a, b):
+        # kve instead of kv works better for large values of p
+        # and smaller values of sqrt(a^2  - b^2)
+        @np.vectorize
+        def _pdf_single(x, p, a, b):
+            return _stats.genhyperbolic_pdf(x, p, a, b)
+
+        return _pdf_single(x, p, a, b)
+
+    def _cdf(self, x, p, a, b):
+
+        @np.vectorize
+        def _cdf_single(x, p, a, b):
+            user_data = np.array(
+                [p, a, b], float
+                ).ctypes.data_as(ctypes.c_void_p)
+            llc = LowLevelCallable.from_cython(
+                _stats, '_genhyperbolic_pdf', user_data
+                )
+
+            t1 = integrate.quad(llc, -np.inf, x)[0]
+
+            if np.isnan(t1):
+                msg = ("Infinite values encountered in scipy.special.kve. "
+                       "Values replaced by NaN to avoid incorrect results.")
+                warnings.warn(msg, RuntimeWarning)
+
+            return t1
+
+        return _cdf_single(x, p, a, b)
+
+    def _rvs(self, p, a, b, size=None, random_state=None):
+        # note: X = b * V + sqrt(V) * X  has a
+        # generalized hyperbolic distribution
+        # if X is standard normal and V is
+        # geninvgauss(p = p, b = t2, loc = loc, scale = t3)
+        t1 = np.float_power(a, 2) - np.float_power(b, 2)
+        # b in the GIG
+        t2 = np.float_power(t1, 0.5)
+        # scale in the GIG
+        t3 = np.float_power(t1, - 0.5)
+        gig = geninvgauss.rvs(
+            p=p,
+            b=t2,
+            scale=t3,
+            size=size,
+            random_state=random_state
+            )
+        normst = norm.rvs(size=size, random_state=random_state)
+
+        return b * gig + np.sqrt(gig) * normst
+
+    def _stats(self, p, a, b):
+        # https://mpra.ub.uni-muenchen.de/19081/1/MPRA_paper_19081.pdf
+        # https://freidok.uni-freiburg.de/fedora/objects/freidok:7974/datastreams/FILE1/content
+        # standardized moments
+        t1 = np.float_power(a, 2) - np.float_power(b, 2)
+        t1 = np.float_power(t1, 0.5)
+        t2 = np.float_power(1, 2) * np.float_power(t1, - 1)
+        b0, b1, b2, b3, b4 = sc.kv(p + np.linspace(0, 4, 5), t1)
```

Comment:
(Surprisingly, this isn't caught by tests.) This needs to be vectorized.

Otherwise, the _stats implementation seems right. I checked by writing my own very obvious implementation. (Would probably have been better to derive the formulas from the moment generating function using SymPy, but this took less thinking.)

```python3
import numpy as np
from scipy.special import kv as K
from scipy.stats import genhyperbolic
from numpy.testing import assert_allclose

p, a, b, loc, scale = 0.5, 1.5, -0.5, 1.5, 2.3

n = 1  # number of tests
test_scale = 100  # scale of tests
p = np.random.rand(n)*test_scale - 5
a = np.random.rand(n)*test_scale
b = (2*np.random.rand(n)-1)*a
loc = test_scale * np.random.rand(n) - 5
scale = test_scale * np.random.rand(n)

def _scipy_2_reference(p, a, b, loc, scale):
    # https://freidok.uni-freiburg.de/fedora/objects/freidok:7974/datastreams/FILE1/content
    # p.23 of the pdf or p.15 from the document page numbering.
    lmbda = p
    delta = scale
    alpha = a/delta
    beta = b/delta
    mu = loc
    zeta = delta * np.sqrt(alpha**2 - beta**2)
    return lmbda, alpha, beta, delta, mu, zeta

lmbda, alpha, beta, delta, mu, zeta = _scipy_2_reference(p, a, b, loc, scale)

mean = mu + beta*delta**2/zeta * K(lmbda+1, zeta)/K(lmbda, zeta)
var = (delta**2/zeta * K(lmbda+1, zeta) / K(lmbda, zeta) +
       beta**2*delta**4/zeta**2
       * (K(lmbda+2, zeta)/K(lmbda, zeta)
          - K(lmbda+1, zeta)**2/K(lmbda, zeta)**2))
skew = (var**(-3/2) *
        (beta**3*delta**6/zeta**3 *
         (K(lmbda+3, zeta)/K(lmbda, zeta)
          - 3*K(lmbda+2, zeta) * K(lmbda+1, zeta) / K(lmbda, zeta)**2
          + 2*K(lmbda+1, zeta)**3 / K(lmbda, zeta)**3)
         + 3*beta*delta**4/zeta**2
         * (K(lmbda+2, zeta)/K(lmbda, zeta)
            - K(lmbda+1, zeta)**2 / K(lmbda, zeta)**2)))
kurtosis = (-3 + var**(-2)
            * (delta**8*beta**4/zeta**4
               * (K(lmbda+4, zeta)/K(lmbda, zeta)
                  - 4*K(lmbda+3, zeta)*K(lmbda+1, zeta)/K(lmbda, zeta)**2
                  + 6*K(lmbda+2, zeta)*K(lmbda+1, zeta)**2/K(lmbda, zeta)**3
                  - 3*K(lmbda+1, zeta)**4/K(lmbda, zeta)**4)
               + delta**6*beta**2/zeta**3
               * (6*K(lmbda+3, zeta)/K(lmbda, zeta)
                  - 12*K(lmbda+2, zeta)*K(lmbda+1, zeta)/K(lmbda, zeta)**2
                  + 6*K(lmbda+1, zeta)**3/K(lmbda, zeta)**3)
               + 3*delta**4/zeta**2
               * K(lmbda+2, zeta)/K(lmbda, zeta)))

mean2 = genhyperbolic.mean(p, a, b, loc, scale)
var2 = genhyperbolic.var(p, a, b, loc, scale)
skew2, kurtosis2 = genhyperbolic.stats(p, a, b, loc, scale, moments="sk")

print(mean, mean2)
print(var, var2)
print(skew, skew2)
print(kurtosis, kurtosis2)

assert_allclose(mean, mean2)
assert_allclose(var, var2)
assert_allclose(skew, skew2)
assert_allclose(kurtosis, kurtosis2)
```

Reply:
Hi @mdhaber - couple of things

1) After change test_moment_r [fails](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=10985&view=logs&j=e54654ce-5799-589e-a4f6-899a89320b3b&t=7a34d531-46c8-5b1f-8e3e-e5bf1eea121f&l=672) due to array mismatch

```
        vals_R     = array([  2.36848367,   8.47393468,  37.88705027, 205.76608511])
        vals_us    = [array([2.36848367]), array([8.47393468]), array([37.88705027]), array([205.76608511])]
```

Ill need to adapt the test so that `vals_R` have the same shape, not a big deal. but yeah i confirm that stats was failing when vectorized - suprisingly this was not caught in tests indeed.

2) the new [test_distr_param_list](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=10985&view=logs&j=e54654ce-5799-589e-a4f6-899a89320b3b&t=7a34d531-46c8-5b1f-8e3e-e5bf1eea121f&l=674) is also failing. it seems the check expects `genhyperbolic` to be in the `invcont` list as well, which is not. However i see many other distcont not in the "inv" list. Not really sure if/how i should add the distribution there it there and what exactly the list is needed for - is there any documentation to look at? I quickly looked and found nothing. I also grep the test folder and could not find any test for `invdistcont` (only passed to mark.parametrized once)

Reply:
2. What continuous distributions aren't in that list? We wrote the check to make sure they are all there - and i checked myself!
The reason for the list is a test that the `support` method does not give bogus values when parameters are invalid. Please just add a set of parameters to the list that is somehow invalid - b > a, for instance.

Reply:
1. I'm not sure why I suggested the change I did. I don't think that solves the problem, as `p` can be n-d, right? I think the intent was to make the list of integers orthogonal to all existing axes. I'll check this out. 
The easy way would be to use `np.vectorize` like the others, but it be better to natively vectorize the function if we can.

Reply:
2 - sorry: the distribution list was indeed complete - if I recall I well was caught offguard as i expected the list to be in alphabetical order or because some distr names are repeated (for the right reason) so the count did not matched. This is not an issue, as you say.

1 - i think you are right there: no matter how you turn the `np.linspace` array (now (N,1) after your suggestion ), vectorization may break when `p` is not transposed coherently - so the current solution above should be safe only if we always expect `p` to be a (N,1) vector.

As a backup - would any of the below ensure full vectorization safety ? i am being explicit and calling the `sc.kv` function 5 times in return it should work regardless from `p`. The rest of the function seem to be correctly vectorized already

lazy option
```
b0 = sc.kv(p + 0, t1)
b1 = sc.kv(p + 1, t1)
b2 = sc.kv(p + 2, t1)
b3 = sc.kv(p + 3, t1)
b4 = sc.kv(p + 4, t1)
```
slightly less lazy
```
b0, b1, b2, b3, b4 = [sc.kv(p + i, t1) for i in np.linspace(0, 4, 5)] 
```

Reply:
Nevermind - i see you did this and should work just fine, thank you

```
integers = np.linspace(0, 4, 5)
# make integers perpendicular to existing dimensions
integers = integers.reshape(integers.shape + (1,) * p.ndim)
b0, b1, b2, b3, b4 = sc.kv(p + integers, t1)
```